analyzers/src/SonarAnalyzer.Common/RegularExpressions/RegexNode.cs

```diff
@@ -0,0 +1,94 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+#nullable enable
+
+using System.Text.RegularExpressions;
+
+namespace SonarAnalyzer.RegularExpressions;
+
+internal sealed class RegexNode
+{
+    public RegexNode(
+        SyntaxNodeValue<string?> pattern,
+        SyntaxNodeValue<RegexOptions?> options)
+    {
+        Pattern = pattern;
+        Options = options;
+    }
+
+    public SyntaxNodeValue<string?> Pattern { get; }
+    public SyntaxNodeValue<RegexOptions?> Options { get; }
+
+    public static RegexNode? FromCtor<TSyntaxKind>(SyntaxNode node, SemanticModel model, ILanguageFacade<TSyntaxKind> language) where TSyntaxKind : struct =>
+        model.GetSymbolInfo(node).Symbol is IMethodSymbol method
+        && method.ContainingType.Is(KnownType.System_Text_RegularExpressions_Regex)
+        && method.IsConstructor()
+        ? FromSymbol(method, node, model, language)
+        : null;
+
+    public static RegexNode? FromMethod<TSyntaxKind>(SyntaxNode node, SemanticModel model, ILanguageFacade<TSyntaxKind> language) where TSyntaxKind : struct =>
+        language.Syntax.NodeIdentifier(node).GetValueOrDefault().Text is { } name
+        && MatchMethods.Any(x => x.Equals(name, language.NameComparison))
+        && model.GetSymbolInfo(node).Symbol is IMethodSymbol method
```

Comment:
We'll need an UT for nonsatic regex IsMatch and Matches where the value looks like invalid regex and it's fine
```suggestion
        && model.GetSymbolInfo(node).Symbol is IMethodSymbol {IsStatic: true } method
```